File: packages/crawler-py/shroodler/extractors/rate_limit.py
from __future__ import annotations

from urllib.parse import urlparse

from shroodler.models import Finding, Form

_LOGIN_ACTION_HINTS = (
    "login",
    "signin",
    "sign-in",
    "log-in",
    "auth",
    "authenticate",
    "reset-password",
    "forgot-password",
    "password-reset",
)

_LOCKOUT_KEYWORDS = (
    "too many attempts",
    "too many requests",
    "rate limit",
    "rate-limit",
    "try again later",
    "temporarily locked",
    "temporarily blocked",
    "account locked",
    "captcha",
)

DEFAULT_ATTEMPTS = 6


def is_login_shaped(form: Form) -> bool:
    """Heuristic: does this form look like a login/auth/reset endpoint?

    Used to scope --check-rate-limit probing to forms where hammering the
    endpoint is actually meaningful (and where the safety tradeoff of
    sending repeated requests is worth it) rather than every form on site.
    """
    types = {(f.type or "").lower() for f in form.fields}
    if "password" in types:
        return True
    action = (form.action or "").lower()
    return any(hint in action for hint in _LOGIN_ACTION_HINTS)


def _probe_values(form: Form, attempt: int) -> dict[str, str]:
    values: dict[str, str] = {}
    for f in form.fields:
        if not f.name:
            continue
        if (f.type or "").lower() == "password":
            values[f.name] = f"shroodler-rl-probe-wrong-{attempt}"
        else:
            values[f.name] = "shroodler-rl-probe"
    return values


def check_form_rate_limit(
    fetcher,
    action: str,
    form: Form,
    *,
    attempts: int = DEFAULT_ATTEMPTS,
) -> list[Finding]:
    """Fire `attempts` rapid requests at a login-shaped form and flag it if
    nothing in the response stream (status, body) suggests any throttling,
    lockout, or CAPTCHA kicked in.

    This is opt-in (--check-rate-limit) and off by default: it deliberately
    sends multiple bad-credential attempts at a real endpoint, which has
    real consequences (account lockout, alerting, log noise) against a
    production target. Only call this against systems you're authorized to
    load-test this way.
    """
    method = (form.method or "POST").upper()
    statuses: list[int] = []
    bodies: list[str] = []
    for i in range(attempts):
        data = _probe_values(form, i)
        if method == "GET":
            resp = fetcher.request("GET", action)
        else:
            resp = fetcher.post_form(action, data)
        if resp.error:
            return []
        statuses.append(resp.status_code)
        bodies.append(resp.text or "")

    if any(s == 429 for s in statuses):
        return []
    lowered = [b.lower() for b in bodies]
    if any(kw in b for b in lowered for kw in _LOCKOUT_KEYWORDS):
        return []
    if len(set(statuses)) > 1:
        # Status changed partway through (e.g. first attempts 200, later
        # ones 403/423) -- treat that as a throttling/lockout signal.
        return []

    ev = f"{attempts} requests, all status {statuses[0] if statuses else '?'}, no lockout/CAPTCHA signal"
    return [
        Finding(
            id="missing-rate-limit",
            severity="medium",
            category="auth",
            url=action,
            description=(
                f"Sent {attempts} rapid requests to this login/auth-shaped form with "
                "bad credentials and saw no rate-limiting, lockout, or CAPTCHA response -- "
                "the endpoint may be brute-forceable."
            ),
            evidence=ev,
        )
    ]


def check_rate_limits(fetcher, origin: str, pages, *, attempts: int = DEFAULT_ATTEMPTS) -> list[Finding]:
    findings: list[Finding] = []
    seen_actions: set[str] = set()
    for page in pages:
        for form in page.forms:
            if not is_login_shaped(form):
                continue
            action = form.action or page.url
            if action.startswith("/"):
                p = urlparse(page.url)
                action = f"{p.scheme}://{p.netloc}{action}"
            if action in seen_actions:
                continue
            seen_actions.add(action)
            findings.extend(check_form_rate_limit(fetcher, action, form, attempts=attempts))
    return findings

File: packages/crawler-go/internal/crawler/rate_limit.go
package crawler

import (
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"

	"github.com/shroodler/crawler-go/internal/models"
)

var loginActionHints = []string{
	"login", "signin", "sign-in", "log-in", "auth", "authenticate",
	"reset-password", "forgot-password", "password-reset",
}

var lockoutKeywords = []string{
	"too many attempts", "too many requests", "rate limit", "rate-limit",
	"try again later", "temporarily locked", "temporarily blocked",
	"account locked", "captcha",
}

const rateLimitAttempts = 6

// isLoginShaped is a heuristic for whether a form looks like a login/auth
// endpoint -- used to scope --check-rate-limit probing to forms where
// hammering the endpoint with repeated requests is actually meaningful.
func isLoginShaped(f models.Form) bool {
	for _, field := range f.Fields {
		if strings.ToLower(field.Type) == "password" {
			return true
		}
	}
	action := strings.ToLower(f.Action)
	for _, hint := range loginActionHints {
		if strings.Contains(action, hint) {
			return true
		}
	}
	return false
}

func rateLimitProbeValues(f models.Form, attempt int) url.Values {
	vals := url.Values{}
	for _, field := range f.Fields {
		if field.Name == "" {
			continue
		}
		if strings.ToLower(field.Type) == "password" {
			vals.Set(field.Name, "shroodler-rl-probe-wrong-"+strconv.Itoa(attempt))
		} else {
			vals.Set(field.Name, "shroodler-rl-probe")
		}
	}
	return vals
}

func doPostForm(client *http.Client, raw string, vals url.Values) fetchResult {
	req, err := http.NewRequest(http.MethodPost, raw, strings.NewReader(vals.Encode()))
	if err != nil {
		return fetchResult{URL: raw}
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	return doRequestObj(client, req, raw)
}

func doRequestObj(client *http.Client, req *http.Request, raw string) fetchResult {
	req.Header.Set("User-Agent", "Shroodler/0.1.0 (+https://shroodler.local)")
	resp, err := client.Do(req)
	if err != nil {
		return fetchResult{URL: raw}
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(io.LimitReader(resp.Body, 2_000_000))
	hdrs := map[string]string{}
	for k, vs := range resp.Header {
		if len(vs) > 0 {
			hdrs[k] = vs[0]
		}
	}
	return fetchResult{URL: raw, Status: resp.StatusCode, Headers: hdrs, Body: string(b), SetCookies: resp.Header.Values("Set-Cookie")}
}

// checkFormRateLimit fires rateLimitAttempts rapid bad-credential requests
// at a login-shaped form and flags it if nothing in the response stream
// (status, body) suggests any throttling, lockout, or CAPTCHA kicked in.
//
// This is opt-in (--check-rate-limit) and off by default: it deliberately
// sends multiple bad-credential attempts at a real endpoint, which has real
// consequences (account lockout, alerting, log noise) against a production
// target. Only call this against systems you're authorized to load-test.
func checkFormRateLimit(client *http.Client, action string, form models.Form) []models.Finding {
	method := strings.ToUpper(form.Method)
	if method == "" {
		method = "POST"
	}
	statuses := make([]int, 0, rateLimitAttempts)
	bodies := make([]string, 0, rateLimitAttempts)
	for i := 0; i < rateLimitAttempts; i++ {
		var res fetchResult
		if method == "GET" {
			res = doGet(client, action, "")
		} else {
			res = doPostForm(client, action, rateLimitProbeValues(form, i))
		}
		if res.Status == 0 {
			return nil
		}
		statuses = append(statuses, res.Status)
		bodies = append(bodies, strings.ToLower(res.Body))
	}
	for _, s := range statuses {
		if s == 429 {
			return nil
		}
	}
	for _, b := range bodies {
		for _, kw := range lockoutKeywords {
			if strings.Contains(b, kw) {
				return nil
			}
		}
	}
	first := statuses[0]
	for _, s := range statuses[1:] {
		if s != first {
			return nil
		}
	}
	ev := "6 requests, all status " + strconv.Itoa(first) + ", no lockout/CAPTCHA signal"
	desc := "Sent 6 rapid requests to this login/auth-shaped form with bad " +
		"credentials and saw no rate-limiting, lockout, or CAPTCHA response -- " +
		"the endpoint may be brute-forceable."
	return []models.Finding{{
		ID: "missing-rate-limit", Severity: "medium", Category: "auth",
		URL: action, Description: desc, Evidence: &ev,
	}}
}

// checkRateLimits scans every discovered page for login-shaped forms and
// probes each distinct action URL once.
func checkRateLimits(client *http.Client, origin string, pages []models.Page) []models.Finding {
	var out []models.Finding
	seen := map[string]bool{}
	for _, page := range pages {
		for _, form := range page.Forms {
			if !isLoginShaped(form) {
				continue
			}
			action := form.Action
			if action == "" {
				action = page.URL
			}
			if strings.HasPrefix(action, "/") {
				if u, err := url.Parse(page.URL); err == nil {
					action = u.Scheme + "://" + u.Host + action
				}
			}
			if seen[action] {
				continue
			}
			seen[action] = true
			out = append(out, checkFormRateLimit(client, action, form)...)
		}
	}
	return out
}
